File: python/parmana/api/policy_api.py
"""
Parmana Policy API.

Validate Runtime policies.
"""

from __future__ import annotations

from typing import Any, cast

from parmana.config.transport import Transport


class PolicyApi:
    """
    Policy API.

    Responsibilities
    ----------------
    - Confirm a policy (name + version) is loadable by the Runtime

    This API does NOT:
    - validate an arbitrary policy document's contents. POST
      /policies/validate (packages/api/src/routes/policies.ts) only
      checks that `policyRepository.load(policyId, policyVersion)`
      succeeds -- it does not accept or check a policy document body.
    - execute Business Transactions
    - verify trust records
    - replay executions
    - generate receipts
    """

    def __init__(
        self,
        transport: Transport,
    ) -> None:
        self._transport = transport

    def validate(
        self,
        policy_id: str,
        policy_version: str,
    ) -> dict[str, Any]:
        """
        Confirm that a policy (name + version) is loadable.

        Parameters
        ----------
        policy_id:
            Policy identifier (`policyId`).

        policy_version:
            Policy version (`policyVersion`).

        Returns
        -------
        `{"valid": bool, "errors": list[str]}`.
        """

        return cast(
            "dict[str, Any]",
            self._transport.send(
                method="POST",
                path="/policies/validate",
                body={
                    "policyId": policy_id,
                    "policyVersion": policy_version,
                },
            ),
        )

File: typescript/src/client/PolicyApi.ts
/**
 * Parmana SDK
 *
 * Policy API.
 *
 * Provides access to policy-related Runtime APIs.
 *
 * Responsibilities:
 * - Validate policies.
 * - Retrieve policy information.
 *
 * This API does NOT:
 * - execute transactions
 * - authorize execution
 * - verify trust records
 * - replay executions
 */

import type {
  Policy,
} from "@parmana/policy";

import type {
  Transport,
} from "../config/Transport.js";

/**
 * Canonical policy validation result.
 */
export interface PolicyValidationResult {
  /**
   * Indicates whether the policy is valid.
   */
  readonly valid: boolean;

  /**
   * Validation errors.
   */
  readonly errors: readonly string[];
}

/**
 * Policy API.
 */
export class PolicyApi {
  constructor(
    private readonly transport: Transport,
  ) {}

  /**
   * Validates a policy.
   */
  public async validate(
    policy: Policy,
  ): Promise<PolicyValidationResult> {
    const response =
      await this.transport.send<PolicyValidationResult>({
        path: "/policies/validate",

        method: "POST",

        body: policy,
      });

    return response.body;
  }
}